packages/react-devtools-shared/src/devtools/views/Profiler/PerfInsight.js

```diff
@@ -0,0 +1,158 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import * as React from 'react';
+import {useContext} from 'react';
+import {enableProfilerPerfInsights} from 'react-devtools-feature-flags';
+import {ProfilerContext} from '../Profiler/ProfilerContext';
+import {StoreContext} from '../context';
+
+import styles from './PerfInsight.css';
+
+type Props = {|
+  fiberID: number,
+|};
+
+export default function PerfInsight({fiberID}: Props) {
+  const {profilerStore} = useContext(StoreContext);
+  const {rootID, selectedCommitIndex} = useContext(ProfilerContext);
+
+  if (!enableProfilerPerfInsights) {
+    return null;
+  }
+
+  // TRICKY
+  // Handle edge case where no commit is selected because of a min-duration filter update.
+  // If the commit index is null, suspending for data below would throw an error.
+  // TODO (ProfilerContext) This check should not be necessary.
+  if (selectedCommitIndex === null) {
+    return null;
+  }
+
+  const {perfInsights} = profilerStore.getCommitData(
+    ((rootID: any): number),
+    selectedCommitIndex,
+  );
+
+  if (perfInsights === null) {
+    return null;
+  }
+
+  const perfInsight = perfInsights.get(fiberID);
+  if (perfInsight == null) {
+    return null;
+  }
+
+  const {
+    isFirstMount,
+    didContextChange,
+    didContextDeepChange,
+    didHooksChange,
+    didHooksDeeplyChange,
+    hooksNeedingMemoization,
+    didPropsChange,
+    didPropsDeepChange,
+    propsNeedingMemoization,
+    didStateChange,
+    didStateDeepChange,
+  } = perfInsight;
+
+  if (isFirstMount) {
+    return null;
+  }
+
+  const insights = [];
+
+  if (!didContextChange && didContextDeepChange) {
+    insights.push(
+      <div key="context" className={styles.Item}>
+        • Context deep value changed
+      </div>,
+    );
+  }
+
+  // If context and hooks didn't change and  props/state are shallow equal,
+  // then this component needs basic memo.
+  if (
+    !didContextChange &&
+    !didHooksChange &&
+    !didPropsChange &&
+    !didStateChange
+  ) {
+    insights.push(
+      <div className={styles.Item}>
+        • Use React.PureComponent or React.Memo to avoid this component
+        rendering.
+      </div>,
+    );
+  }
+
+  if (hooksNeedingMemoization != null && hooksNeedingMemoization.length > 0) {
+    const hooksList = hooksNeedingMemoization.map(key => (
+      <span key={key} className={styles.Key}>
+        {key}
+      </span>
+    ));
+    insights.push(
+      <div className={styles.Item}>
+        {didHooksDeeplyChange
+          ? '• Some hooks could avoid updating by using memoization:'
+          : '• This component could avoid updating by using memoization for hooks:'}
```

Comment:
I don't think the distinction between these two is clear. Seems like both are saying the same thing but with slightly different words.

Reply:
Tried to make this more explict.